File: server/src/providers/registry.ts
/**
 * @fileoverview Provider Registry & Factory
 *
 * Centralized configuration for all external service providers.
 * Factory functions to create provider instances based on environment.
 *
 * @module providers/registry
 */

import type { RealEstateProvider } from './real-estate/types.js'
import type { LLMProvider } from './llm/types.js'

/**
 * Provider configuration metadata
 */
export interface ProviderConfig {
  name: string
  envKeys: readonly string[]
  baseUrl?: string
  defaultModel?: string
  rateLimit?: {
    requests: number
    period: 'minute' | 'hour' | 'day' | 'month'
  }
}

/**
 * Registry of all available providers and their configuration
 */
export const PROVIDER_REGISTRY = {
  realEstate: {
    rapidapi_zillow: {
      name: 'RapidAPI Zillow',
      envKeys: ['RAPIDAPI_KEY'],
      baseUrl: 'https://zillow-com1.p.rapidapi.com',
      rateLimit: { requests: 100, period: 'month' as const },
    },
    simplyrets: {
      name: 'SimplyRETS',
      envKeys: ['SIMPLYRETS_API_KEY', 'SIMPLYRETS_API_SECRET'],
      baseUrl: 'https://api.simplyrets.com',
      rateLimit: { requests: 1000, period: 'day' as const },
    },
    mock: {
      name: 'Mock Provider',
      envKeys: [],
      baseUrl: undefined,
      rateLimit: undefined,
    },
  },
  llm: {
    openai: {
      name: 'OpenAI',
      envKeys: ['OPENAI_API_KEY'],
      baseUrl: 'https://api.openai.com/v1',
      defaultModel: 'gpt-4o-mini',
    },
    anthropic: {
      name: 'Anthropic',
      envKeys: ['ANTHROPIC_API_KEY'],
      baseUrl: 'https://api.anthropic.com/v1',
      defaultModel: 'claude-3-haiku-20240307',
    },
    ollama: {
      name: 'Ollama (Local)',
      envKeys: [],
      baseUrl: process.env.OLLAMA_BASE_URL || 'http://localhost:11434',
      defaultModel: process.env.OLLAMA_MODEL || 'llama3.2',
    },
    mock: {
      name: 'Mock Provider',
      envKeys: [],
      baseUrl: undefined,
      defaultModel: 'mock-model',
    },
  },
} as const

/**
 * Get the configured real estate provider type from environment
 */
export function getRealEstateProviderType(): keyof typeof PROVIDER_REGISTRY.realEstate {
  const provider = process.env.REAL_ESTATE_PROVIDER || 'mock'
  if (provider in PROVIDER_REGISTRY.realEstate) {
    return provider as keyof typeof PROVIDER_REGISTRY.realEstate
  }
  console.warn(`Unknown real estate provider: ${provider}, falling back to mock`)
  return 'mock'
}

/**
 * Get the configured LLM provider type from environment
 */
export function getLLMProviderType(): keyof typeof PROVIDER_REGISTRY.llm {
  const provider = process.env.LLM_PROVIDER || 'mock'
  if (provider in PROVIDER_REGISTRY.llm) {
    return provider as keyof typeof PROVIDER_REGISTRY.llm
  }
  console.warn(`Unknown LLM provider: ${provider}, falling back to mock`)
  return 'mock'
}

/**
 * Check if required environment variables are set for a provider
 */
export function checkProviderEnv(config: ProviderConfig): { valid: boolean; missing: string[] } {
  const missing = config.envKeys.filter(key => !process.env[key])
  return {
    valid: missing.length === 0,
    missing,
  }
}

// Lazy-loaded provider instances (singleton pattern)
let realEstateProviderInstance: RealEstateProvider | null = null
let llmProviderInstance: LLMProvider | null = null

/**
 * Create or get the real estate provider instance
 */
export async function createRealEstateProvider(): Promise<RealEstateProvider> {
  if (realEstateProviderInstance) {
    return realEstateProviderInstance
  }

  const providerType = getRealEstateProviderType()
  const config = PROVIDER_REGISTRY.realEstate[providerType]

  // Check environment variables
  const envCheck = checkProviderEnv(config)
  if (!envCheck.valid && providerType !== 'mock') {
    console.warn(
      `Missing env vars for ${config.name}: ${envCheck.missing.join(', ')}. Falling back to mock provider.`
    )
    const { MockRealEstateProvider } = await import('./real-estate/mock.js')
    realEstateProviderInstance = new MockRealEstateProvider()
    return realEstateProviderInstance
  }

  // Import and instantiate the appropriate provider
  switch (providerType) {
    case 'rapidapi_zillow': {
      const { RapidAPIZillowProvider } = await import('./real-estate/rapidapi-zillow.js')
      realEstateProviderInstance = new RapidAPIZillowProvider()
      break
    }
    case 'simplyrets': {
      const { SimplyRetsProvider } = await import('./real-estate/simplyrets.js')
      realEstateProviderInstance = new SimplyRetsProvider()
      break
    }
    case 'mock':
    default: {
      const { MockRealEstateProvider } = await import('./real-estate/mock.js')
      realEstateProviderInstance = new MockRealEstateProvider()
      break
    }
  }

  console.log(`Real estate provider initialized: ${realEstateProviderInstance!.name}`)
  return realEstateProviderInstance!
}

/**
 * Create or get the LLM provider instance
 */
export async function createLLMProvider(): Promise<LLMProvider> {
  if (llmProviderInstance) {
    return llmProviderInstance
  }

  const providerType = getLLMProviderType()
  const config = PROVIDER_REGISTRY.llm[providerType]

  // Check environment variables
  const envCheck = checkProviderEnv(config)
  if (!envCheck.valid && providerType !== 'mock' && providerType !== 'ollama') {
    console.warn(
      `Missing env vars for ${config.name}: ${envCheck.missing.join(', ')}. Falling back to mock provider.`
    )
    const { MockLLMProvider } = await import('./llm/mock.js')
    llmProviderInstance = new MockLLMProvider()
    return llmProviderInstance
  }

  // Import and instantiate the appropriate provider
  switch (providerType) {
    case 'openai': {
      const { OpenAIProvider } = await import('./llm/openai.js')
      llmProviderInstance = new OpenAIProvider()
      break
    }
    case 'anthropic': {
      const { AnthropicProvider } = await import('./llm/anthropic.js')
      llmProviderInstance = new AnthropicProvider()
      break
    }
    case 'ollama': {
      const { OllamaProvider } = await import('./llm/ollama.js')
      llmProviderInstance = new OllamaProvider()
      break
    }
    case 'mock':
    default: {
      const { MockLLMProvider } = await import('./llm/mock.js')
      llmProviderInstance = new MockLLMProvider()
      break
    }
  }

  console.log(`LLM provider initialized: ${llmProviderInstance!.name}`)
  return llmProviderInstance!
}

/**
 * Reset provider instances (useful for testing)
 */
export function resetProviders(): void {
  realEstateProviderInstance = null
  llmProviderInstance = null
}

/**
 * Get status of all configured providers
 */
export async function getProvidersStatus(): Promise<{
  realEstate: { provider: string; available: boolean; config: ProviderConfig }
  llm: { provider: string; available: boolean; config: ProviderConfig }
}> {
  const realEstateType = getRealEstateProviderType()
  const llmType = getLLMProviderType()

  const realEstateProvider = await createRealEstateProvider()
  const llmProvider = await createLLMProvider()

  return {
    realEstate: {
      provider: realEstateType,
      available: await realEstateProvider.isAvailable(),
      config: PROVIDER_REGISTRY.realEstate[realEstateType],
    },
    llm: {
      provider: llmType,
      available: await llmProvider.isAvailable(),
      config: PROVIDER_REGISTRY.llm[llmType],
    },
  }
}

File: ai-service/src/providers/registry.py
"""
LLM Provider Registry and Factory.

Central configuration for all available LLM providers.
Provider selection via LLM_PROVIDER environment variable.
"""

import os
from typing import Dict, Any, Optional
from .types import LLMProvider
from .ollama import OllamaProvider
from .openai_provider import OpenAIProvider
from .mock import MockLLMProvider


# Provider configuration registry
PROVIDER_CONFIG: Dict[str, Dict[str, Any]] = {
    "ollama": {
        "name": "Ollama (Local)",
        "description": "Run AI locally on your machine - free, private",
        "env_keys": [],  # No API key needed
        "default_model": "llama3.2",
        "setup_url": "https://ollama.ai",
        "setup_steps": [
            "Install Ollama from https://ollama.ai",
            "Run: ollama pull llama3.2",
            "Ollama runs automatically after install"
        ]
    },
    "openai": {
        "name": "OpenAI",
        "description": "Cloud-based AI with high accuracy",
        "env_keys": ["OPENAI_API_KEY"],
        "default_model": "gpt-4o-mini",
        "setup_url": "https://platform.openai.com/api-keys",
        "setup_steps": [
            "Create account at https://platform.openai.com",
            "Generate API key",
            "Add OPENAI_API_KEY to .env"
        ]
    },
    "mock": {
        "name": "Mock Provider",
        "description": "Fake responses for testing - always available",
        "env_keys": [],
        "default_model": "mock-v1",
        "setup_url": None,
        "setup_steps": ["No setup required"]
    }
}


def get_available_providers() -> Dict[str, Dict[str, Any]]:
    """Get information about all available providers."""
    return PROVIDER_CONFIG.copy()


def create_llm_provider(provider_name: Optional[str] = None) -> LLMProvider:
    """
    Create an LLM provider instance.

    Args:
        provider_name: Provider to use. If None, reads from LLM_PROVIDER env var.
                      Falls back to 'ollama', then 'mock' if not available.

    Returns:
        LLMProvider instance
    """
    # Determine provider
    if provider_name is None:
        provider_name = os.getenv("LLM_PROVIDER", "ollama")

    provider_name = provider_name.lower()

    # Create provider instance
    if provider_name == "ollama":
        return OllamaProvider()
    elif provider_name == "openai":
        return OpenAIProvider()
    elif provider_name == "mock":
        return MockLLMProvider()
    else:
        print(f"Unknown provider '{provider_name}', falling back to mock")
        return MockLLMProvider()


async def get_best_available_provider() -> LLMProvider:
    """
    Get the best available LLM provider.

    Checks providers in order of preference:
    1. Configured provider (LLM_PROVIDER env var)
    2. Ollama (if running)
    3. OpenAI (if key configured)
    4. Mock (always available)

    Returns:
        The best available LLMProvider
    """
    # First try the configured provider
    configured = os.getenv("LLM_PROVIDER", "").lower()
    if configured:
        provider = create_llm_provider(configured)
        if await provider.is_available():
            return provider
        print(f"Configured provider '{configured}' not available")

    # Try Ollama
    ollama = OllamaProvider()
    if await ollama.is_available():
        print("Using Ollama for LLM inference")
        return ollama

    # Try OpenAI
    openai = OpenAIProvider()
    if await openai.is_available():
        print("Using OpenAI for LLM inference")
        return openai

    # Fall back to mock
    print("No LLM provider available, using mock")
    return MockLLMProvider()
